### PR TITLE
Feature: ActionBar

### LINKED ISSUE #1790: Add ActionBar component
image

# 🙋 Feature Request

An ActionBar allows users to perform bulk actions on the selected items in collection components like TableView, GridView, ColumnView, etc. It composes an [ActionGroup](https://react-spectrum.adobe.com/react-spectrum/ActionGroup.html) with a count of selected items and a button to clear the selection.

## 💁 API

The action bar will be split into two components: the `ActionBar` itself, and an `ActionBarContainer` (name TBD), which provides the layout for a collection component (e.g. Table) and the bar.

The `ActionBar` is a collection component. It accepts `<Item>` elements as children just like an `ActionGroup`. It should be implemented with `ActionGroup` internally, but only expose a subset of options.

```ts
interface ActionBarProps extends CollectionBase {
  // Required. When zero, the ActionBar is hidden.
  selectedItemCount: number,
  // Also required. When clicked, the selection should be cleared.
  onClearSelection: () => void,
  isEmphasized?: boolean,
  onAction?: (key: Key) => void
}
```

The `ActionBarContainer` adjusts the size of the collection component when it becomes visible to reduce the size such that the ActionBar does not cover the content and the user can still scroll to the last row. This can likely be implemented using [slots](https://react-spectrum.adobe.com/react-spectrum/layout.html#slots).

```ts
interface ActionBarContainer {
  children: ReactNode
}
```

### Example

```jsx

<ActionBarContainer>
  <Table selectedKeys={selectedKeys}>
    {...}
  </Table>
  <ActionBar
    selectedItemCount={selectedKeys.size}
    onClearSelection={...}
    isEmphasized>
    <Item>Edit</Item>
    <Item>Delete</Item>
  </ActionBar>
</ActionBarContainer>
```

### Questions for design
- What ActionGroup properties should be supported?
    - ~~Justified?~~ not for now
    - ~~Density?~~ not for now
    - Is it always quiet?
    - I assume not emphasized because the bar also has a different emphasis?
    - Not orientation
    - Probably not selection?

## 🧢 Your Company/Team

RSP

### PR DESCRIPTION
**Pull request by iamstiil**
Closes #1790

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:
// -- //

## 🧢 Your Project:
// -- //

Hey,
I started working on the ActionBar component. Hope that is ok 😄 
I don't really know if the way went with it is fine.
How should I proceed?

### CODE REVIEW

**Comment by snowystinger** [comment]
Hey! Thanks so much for the PR, someone from the team will take a look at this tomorrow.

File: packages/@react-spectrum/actionbar/src/ActionBarContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import {classNames, useDOMRef, useStyleProps} from '@react-spectrum/utils';
+import {DOMRef} from '@react-types/shared';
+import {filterDOMProps} from '@react-aria/utils';
+import React from 'react';
+import {SpectrumActionBarContainerProps} from '@react-types/actionbar';
+import {useProviderProps} from '@react-spectrum/provider';
+
+
+function ActionBarContainer(props: SpectrumActionBarContainerProps, ref: DOMRef<HTMLDivElement>) {
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/src/ActionBarContainer.tsx:21]: So my understanding of this component based on the issue is that even if the Table is given a height, the ActionBarContainer height should not change.
As such, we may need for the final dom to look something like this
```
<ActionBarContainer> -> display: flex, flexDirection: column
  <Table/> -> flex: 1 1 auto;
  <ActionBar/> -> flex: 0 0 auto;
</ActionBarContainer>
```

To automatically get those styles set on the components inside the ActionBarContainer, we may need to use Slots. There are examples of Slots usage in Dialog https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/dialog/src/Dialog.tsx#L97 as well as docs discussing usage of already implemented ones https://react-spectrum.adobe.com/react-spectrum/layout.html#slots

File: packages/@react-spectrum/actionbar/stories/ActionBar.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            </TableHeader>
+            <TableBody items={items}>
+              {item =>
+                (<Row key={item.foo}>
+                  {key => <Cell>{item[key]}</Cell>}
+                </Row>)
+              }
+            </TableBody>
+          </Table>
+          <ActionBar
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/stories/ActionBar.stories.tsx:60]: currently the ActionBar isn't hiding when there are 0 selected

File: packages/@react-spectrum/actionbar/stories/ActionBar.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            selectedItemCount={selectedKeys === 'all' ? selectedKeys : selectedKeys.size}
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/stories/ActionBar.stories.tsx:61]: your types say that selectedItemCount is a number, yet you pass it a string if it's 'all'
I don't see where you handle 'all' anywhere else

File: packages/@react-spectrum/actionbar/docs/ActionBar.mdx [hunk 4]
```diff
@@ -0,0 +1,72 @@
+<!-- Copyright 2020 Adobe. All rights reserved.
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/docs/ActionBar.mdx:1]: lets delete this file for the moment

**Review by snowystinger — COMMENTED** [review]
Hey there, thanks again for starting this. I've taken a quick glance over it.
Other things to note, this is currently failing lint and tests, you can run those locally.

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  return (
+    <div
+      {...filterDOMProps(props)}
+      {...styleProps}
+      ref={domRef}
+      className={classNames(
+        styles,
+        'spectrum-ActionBar', {
+          'spectrum-ActionBar--primary': variant === 'primary',
```
[the remark this answers is not in the document]
Comment (iamstiil) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:52]: I just thought that, since the wireframe had a light background with dark text and the design had the primary color with white text, there might be variants involved I'm currently unaware of. So I already added the logic for those while testing out stuff :) You should just tell me if you want it to be left there or removed 😄

File: packages/@react-spectrum/actionbar/src/actionbar.css [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  &.spectrum-ActionBar--secondary {
+    background-color: var(--spectrum-actionbar-secondary-background-color);
+  }
+
+  &.spectrum-ActionBar--warning {
+    background-color: var(--spectrum-actionbar-warning-background-color);
+  }
+
+  .spectrum-ActionBar__leading {
```
[the remark this answers is not in the document]
Comment (iamstiil) [reply packages/@react-spectrum/actionbar/src/actionbar.css:30]: `react-spectrum-ActionBar--leading` is the class targeting the left side of actionbar. This class goes with the div containing the close button and the text. Since it looked like you peepz use BEM and that div represents an element in the component, I just went with the double underscore notation. Maybe, since you know the context for the class, you might have a better name?

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        'react-spectrum-ActionBar', {
+          'react-spectrum-ActionBar--primary': variant === 'primary',
+          'react-spectrum-ActionBar--secondary': variant === 'secondary',
+          'react-spectrum-ActionBar--warning': variant === 'warning',
+          'react-spectrum-ActionBar--emphasized': isEmphasized
+        },
+        styleProps.className
+      )}>
+      <div ref={innerDivRef}>
+        <Provider {...providerProps}>
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/src/ActionBar.tsx:67]: we should add `isDisabled` when selecteditemCount is 0, that way we don't have extra tab stops when it's hidden (should also be fixed by another PR that is adjusting focus management, but this is probably the correct thing to do anyways short of completely un-rendering them after they transition out)

File: packages/@react-spectrum/actionbar/src/actionbar.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/actionbar/src/actionbar.css]: ah, then it's just `react-spectrum-ActionBar-leading`
we almost use BEM

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx]: Ah, yep, no variants planned at this time, only emphasized to change to the blue background

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    isEmphasized,
+    onAction,
+    onClearSelection,
+    selectedItemCount,
+    variant,
+    ...otherProps
+  } = props;
+
+  const innerDivRef = React.useRef<HTMLDivElement>();
+  const height = innerDivRef.current?.clientHeight;
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/src/ActionBar.tsx:43]: this mostly works, but it still causes the entire ActionBarContainer to change size, the table should shrink by the amount that the ActionBar grows
also, best to use a `useLayoutEffect` and not do it directly in the render

this is more of a note to the rest of the team to get some opinions:
I know it should probably always be the same height, but on the off change there's a really long string and small width table, we may also need to listen to size changes, should we use https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts or should this be impossible?

**Review by snowystinger — COMMENTED** [review]
thanks for the prompt changes!

**Review by devongovett — APPROVED** [review]
